### PR TITLE
Fix: Label over hovered building not shown for height Metric value of…

### PR DESCRIPTION
**Pull request by Shivan-997**
… zero #1623

# Hovering some leaves does not show a temporary label

_**Please read the [CONTRIBUTING.md](https://github.com/MaibornWolff/codecharta/blob/main/CONTRIBUTING.md) before opening a PR.**_

Closes: #1623

## Description
codeMapLabelService should add label even if node has a height attribute value of 0.
Notes: There is no Bug in Multiple Mode because we have found that in multiple mode there must be common metrics in order to be visualize the multiple "maps".

## Screenshots or gifs
Screenshot 2021-01-22 at 13 24 09

### CODE REVIEW

**Review by ce-bo — APPROVED** [review]
LGTM

File: visualization/package.json [hunk 1]
```diff
@@ -136,6 +136,7 @@
 		"angularjs-color-picker": "^3.4.8",
 		"babel-loader": "^8.1.0",
 		"bestzip": "^2.1.5",
+		"canvas": "^2.6.1",
```
Comment (BridgeAR) [inline visualization/package.json:139]: Is it intentional that this is added? I can't find any usage in this code?

File: visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts [hunk 2]
```diff
@@ -62,51 +62,49 @@ export class CodeMapLabelService implements CameraChangeSubscriber {
...
-				}
-				labelText += `${node.attributes[state.dynamicSettings.heightMetric]} ${state.dynamicSettings.heightMetric}`
+		if (options.showNodeName) {
+			labelText = `${node.name}`
+		}
+		if (options.showNodeMetric) {
+			if (labelText !== "") {
+				labelText += "\n"
 			}
+			labelText += `${node.attributes[state.dynamicSettings.heightMetric]} ${state.dynamicSettings.heightMetric}`
```
Comment (BridgeAR) [inline visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts:74]: Are we safe that `node.attributes` is indeed an object? We originally had a safeguard at the top (`node.attributes?.`).

File: visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts [hunk 3]
```diff
@@ -62,51 +62,49 @@ export class CodeMapLabelService implements CameraChangeSubscriber {
...
+		const {
+			appSettings: { layoutAlgorithm }
+		} = state
+		const labelHeightScaled = this.LABEL_HEIGHT_COEFFICIENT * margin * this.LABEL_SCALE_FACTOR
+		let labelOffset = labelHeightScaled + label.heightValue / 2
+
+		switch (layoutAlgorithm) {
+			// !remark : algorithm scaling is not same as the squarified layout,
+			// !layout offset needs to be scaled down,the divided by value is just empirical,
+			// TODO !needs further investigation
```
Comment (BridgeAR) [inline visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts:87]: We might find a nicer description here but I can't think of anything right now :)

File: visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts [hunk 4]
```diff
@@ -222,14 +220,16 @@ export class CodeMapLabelService implements CameraChangeSubscriber {
...
 		const mapCenter = new Box3().setFromObject(this.threeSceneService.mapGeometry).getBoundingSphere(new Sphere()).center
-		const distance = this.threeCameraService.camera.position.distanceTo(mapCenter)
-		if (label !== null) {
-			this.lineCount = label.lineCount
-		}
-		if (this.lineCount > 1) {
-			sprite.scale.set((distance / this.LABEL_WIDTH_DIVISOR) * labelWidth, distance / 25, 1)
-		} else {
-			sprite.scale.set((distance / this.LABEL_WIDTH_DIVISOR) * labelWidth, distance / this.LABEL_HEIGHT_DIVISOR, 1)
+		if (this.threeCameraService.camera) {
```
Comment (BridgeAR) [inline visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts:223]: Was this check not necessary before?

**Review by BridgeAR — APPROVED** [review]
LGTM % comments

**Pull request merged**